### ISSUE TITLE
Memory reporting is inconsistent with "free"

### ISSUE THREAD

**Issue by hholst80** [dylanaraps/neofetch#2418]
## Description

The total memory amount is consistent but the free amount does not add up to what 'free' reports using the same units.  I suspect that this is due to the handling of 'shared memory' that simply should be ignored, and is accounted for in the other fields already.

The reference to how the mess in /proc/meminfo should be interpreted is undoubtedly 'free.c' from procps.

https://gitlab.com/procps-ng/procps/-/tree/master?ref_type=heads

#### Neofetch version

7.1.0

## Screenshot

![image](https://github.com/dylanaraps/neofetch/assets/6200749/5000de9f-1303-47ad-aa95-fcec7fed4722)

## Config file

nil

## Verbose log

https://pastebin.com/GAcSZZWc

**Comment by hholst80** [comment]
free is consistent with what the kernel developers suggest user space to do: 

    MemTotal = MemAvailable + "Used" 


where "used" is non-volatile, allocated memory in various ways, but not directly defined by /proc/meminfo.  With basic algebra we can derive the total lump of "used" memory from the total and the kernel reported available memory: 

    "Used" = MemTotal - MemAvailable

This is what free effectively does:

![image](https://github.com/dylanaraps/neofetch/assets/6200749/3bea6edc-863c-4ee4-8676-173cd8426a8f)

We do not need to break any 2.6 kernel users out there we can simply check the kernel for existance of 'MemoryAvailable'.  It is important to get these things right because other tools look at random other tools and errors propagate in the ecosystem.  Just look at the number of references to this issue: https://github.com/KittyKatt/screenFetch/issues/386#issuecomment-249312716

We can happily leave the taxonomy details of how memory is micro managed by the kernel to the kernel API, and achieve less work, and consistency with the reference report by 'free'.

I patched my neofetch locally and with a green light I can submit a PR to share it.

![image](https://github.com/dylanaraps/neofetch/assets/6200749/0643f053-5113-43c1-8394-e159a1e2021f)

**Comment by hholst80** [comment]
@tobbez Torbjörn, do you care to comment on this things or have you moved on to more important things than bit-picking on memory reporting? ;-)

**Comment by tobbez** [comment]
Using `MemTotal - MemAvailable` would be reasonable, and that's also what procps-ng does (falling back to `MemTotal - MemFree` if the result is negative).

Problem is, neofetch already uses `MemAvailable` if it is available (since 106a53c575d6dc97ed461139c0fd22991a5528f1):

https://github.com/dylanaraps/neofetch/blob/ccd5d9f52609bbdcd5d8fa78c4fdb0f12954125f/neofetch#L2689-L2701

In addition, the handling of Shmem in the fallback path should still be correct.

The logic behind including `Shmem` becomes clearer if you look at the [conky code from which it originates](https://github.com/brndnmtthws/conky/blob/c43582e917fa124a2a9bc3a8224f74b2bf4b4160/src/linux.cc#L229-L234):

```c
  /* Reclaimable memory: does not include shared memory, which is part of cached
     but unreclaimable. Includes the reclaimable part of the Slab cache though.
     Note: when shared memory is swapped out, shmem decreases and swapfree
     decreases - we want this.
  */
  curbufmem = (info.cached - info.shmem) + info.buffers + sreclaimable;
```

So the calculation currently used:

```
used = MemTotal + Shmem - MemFree - Buffers - Cached - SReclaimable
```

was simplified from:

```
used = MemTotal - MemFree - Buffers - (Cached - Shmem) - SReclaimable
```

That `Shmem` should indeed be excluded from `Cached` is confirmed by posts on the Linux kernel mailing list from 2016 (discussing a patch that proposed excluding shmem and driver pages from Cached in /proc/meminfo, precisely because of this confusion):

> Even before we added MemAvailable, users knew that page cache is
> easily convertible to free memory on pressure, and estimated their
> "available" memory by looking at the sum of MemFree, Cached, Buffers.
> However, "Cached" is calculated using NR_FILE_PAGES, which includes
> shmem and random driver pages inserted into the page tables; neither
> of which are easily reclaimable, or reclaimable at all. Reclaiming
> shmem requires swapping, which is slow. And unlike page cache, which
> has fairly conservative dirty limits, all of shmem needs to be written
> out before becoming evictable. Without swap, shmem is not evictable at
> all. And driver pages certainly never are.
> 
> Calling these pages "Cached" is misleading and has resulted in broken
> formulas in userspace. They misrepresent the memory situation and
> cause either waste or unexpected OOM kills.

-- https://lore.kernel.org/lkml/1455827801-13082-1-git-send-email-hannes@cmpxchg.org/

> We have to assume that Cached has been useful to some people, and that
> they've learnt to subtract Shmem from it, if slow or no swap concerns them.

-- https://lore.kernel.org/lkml/alpine.LSU.2.11.1602181422550.2289@eggly.anvils/

> I think everything will ok. Subtraction of shmem isn't widespread practice,
> more like secret knowledge. This wasn't documented and people who use
> this should be aware that this might stop working at any time. So, ACK.

-- https://lore.kernel.org/lkml/CALYGNiMHAtaZfGovYeud65Eix8v0OSWSx8F=4K+pqF6akQah0A@mail.gmail.com/

(it appears the patch under discussion was not subsequently committed)

**Comment by hholst80** [comment]
Thank you for the input!

> Problem is, neofetch already uses `MemAvailable` if it is available (since [106a53c](https://github.com/dylanaraps/neofetch/commit/106a53c575d6dc97ed461139c0fd22991a5528f1)):

Look at that!  Great to see that I was not alone in thinking that this could be improved.  

No action required here then.  Well, we want that next release soon. 7.1.0 is 2 years old! ;-)
